lms/handle-vitally-sync-rate-limit

anathomical opened the pull request:
## WHAT
Allow SyncOrganization worker to handle rate-limiting from the API
## WHY
Organizations (`District` records in our database) are synced to Vitally via a different API than users or schools.  This different API has rate-limiting built in rate limiting.  Our limited rate is 1000 items per minute, and resets 60 seconds after the first request, but the API is fast and we're currently attempting to sync 12k items in this 60 second window without checking to see if they succeed.
## HOW
The new code examines the result of our API call, and if it is rate limited (as indicated by a 429 response code, according to Vitally documentation: https://docs.vitally.io/pushing-data-to-vitally/rest-api#rate-limiting) we schedule a retry to happen randomly between 2 and 20 minutes in the future.  The two minute floor should ensure we don't retry until after our current rate limit window passes, and allowing reschedules out to 20 minutes should, on average, allow us to submit around 19~20k records that don't trigger a second rate-limit response.  Given that we only have around 12k records to sync at the moment, that should give us plenty of overhead.

### Notion Card Links
https://www.notion.so/quill/Modify-SyncVitallyOrganizationsWorker-to-gracefully-handle-rate-limiting-from-Vitally-s-new-API-fe0fd9f61d26402786109da6aa697bf6

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A
Back-to-school 2022: Have you checked the [webinar schedule](https://www.notion.so/quill/Back-to-school-webinar-banners-2022-a75a89cfad9f434899ef6be3eb184733) to avoid for downtime/risky deploys? | N/A

anathomical committed:
Allow SyncOrganization worker to handle rate-limiting from the API

When we get a rate-limited response, we'll randomly schedule the item to get re-processed in 2 to 20 minutes.  Our limited rate is 1000 items per minute, and resets 60 seconds after the first request, so this should give us room for about 20k items while we currently sync around 12k

anathomical committed:
Fix linting error

happythenewsad commented on services/QuillLMS/app/workers/sync_vitally_organization_worker.rb:
```diff
@@ -3,9 +3,25 @@
 class SyncVitallyOrganizationWorker
```
Looks good. There's a chance that even with this enhancement, jobs will continue to "fail forward" for a long time. We won't be able to diagnose this from the job retry count, since the soft failure path in this PR leads to the creation of a fresh job. 

I'd like to consider:
- widening the requeue time to the max acceptable by stakeholders
- adding New Relic or some other alerting, so that we can track "long-failing" jobs

happythenewsad reviewed (commented):
Very reasonable solution. However, it allows the possibility for jobs to get rolled forward indefinitely due to 429 issues

anathomical committed:
Send an error to NewRelic and Sentry when we hit API rate limits

Don't actually raise the error since we don't want to trigger Sidekiq retry logic, but we do want to report when this happens so we can improve our code.

anathomical committed:
Avoid hitting rate limits when syncing Organization data to Vitally

We were just throwing all of our items at their API at once, but since we know we're limited to 1k requests per minute, we can refactor the job and use perform_in to space out our requests to avoid rate limiting

anathomical committed:
Fix linting errors

happythenewsad reviewed (approved):
Great additions, LGTM!
